### PR TITLE
feat(hydrated_cubit): hydrated storage compaction

### PR DESCRIPTION
**Pull request by orsenkucher**
## Status

**IN DEVELOPMENT** - one test is missing, please help🥺🥺🥺🥺🥺

## Breaking Changes

NO

## Description

With intention to address #1422 issue,
add `compact` method to `HydratedStorage` class, but not `Storage` interface.
So you'll need to downcast.

In addition to `compact`, introduce `compactionStrategy` for `HydratedStorage.build` function.
## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: packages/hydrated_cubit/lib/src/hydrated_storage.dart [hunk 1]
```diff
@@ -83,9 +98,8 @@ class HydratedStorage implements Storage {
             await box.put(key, object);
           } on dynamic catch (_) {}
         }
-      } on dynamic catch (_) {} finally {
-        await file.delete();
-      }
+      } on dynamic catch (_) {}
+      await file.delete();
     }
```
Comment (orsenkucher) [inline packages/hydrated_cubit/lib/src/hydrated_storage.dart:103]: I've also removed `finally` bloc from here

Comment (orsenkucher) [reply packages/hydrated_cubit/lib/src/hydrated_storage.dart:103]: woooops, `finally` block haha!

File: packages/hydrated_cubit/lib/src/hydrated_storage.dart [hunk 2]
```diff
@@ -62,6 +76,7 @@ class HydratedStorage implements Storage {
       final box = await hive.openBox<dynamic>(
         'hydrated_box',
         encryptionCipher: encryptionCipher,
+        compactionStrategy: compactionStrategy ?? defaultCompactionStrategy,
```
Comment (orsenkucher) [inline packages/hydrated_cubit/lib/src/hydrated_storage.dart:79]: In the test with `TODO(need_help)` I actually want to make sure this line is present and does what it should

File: packages/hydrated_cubit/lib/src/hydrated_storage.dart [hunk 3]
```diff
@@ -126,4 +140,16 @@ class HydratedStorage implements Storage {
...
+
+  /// Induces manual box compaction. It's rarely needed.
+  /// Consider using custom compaction strategy instead.
+  ///
+  /// `compact` is [HydratedStorage] specific feature,
+  /// to use it downcast `Storage` to `HydratedStorage`.
+  Future<void> compact() {
+    return _box.isOpen ? _lock.synchronized(_box.compact) : Future.value();
+  }
 }
```
Comment (felangel) [inline packages/hydrated_cubit/lib/src/hydrated_storage.dart:152]: Thoughts on avoiding exposing the `compact` API?

Comment (orsenkucher) [reply packages/hydrated_cubit/lib/src/hydrated_storage.dart:152]: > Thoughts on avoiding exposing the `compact` API?

I think every Bitcask should have compaction api

**Comment by felangel** [comment]
Closing this because hive handles automatically compacting based on the `defaultCompactionStrategy` https://github.com/hivedb/hive/blob/eb300fc1394df1184ddc8b739403b6416a1b57b9/hive/lib/src/box/box_impl.dart#L91

**Comment by karabanovbs** [comment]
> Closing this because hive handles automatically compacting based on the `defaultCompactionStrategy` https://github.com/hivedb/hive/blob/eb300fc1394df1184ddc8b739403b6416a1b57b9/hive/lib/src/box/box_impl.dart#L91

Hi, if i want to use custom `compactionStrategy`, should i implement custom storage?

**Pull request closed without merging**